### PR TITLE
PEP-517 source distribution support

### LINKED ISSUE #573: Add support for PEP-517
Currently `tox` does not yet appear have support for the `buildsystem.requires` key of `pyproject.toml` as defined in [PEP-518](https://www.python.org/dev/peps/pep-0518/).

PIP has implemented this: https://github.com/pypa/pip/pull/4144 (will be part of PIP 10.0) and it would be nice if tox could also add support for installing packages defined using that key, instead of having to use the [broken](https://www.python.org/dev/peps/pep-0518/#rationale) `setup_requires` key of `./setup.py`.

Example of a `pyproject.toml` file that creates a setuptools 34+ environment:

```
[build-system]
requires = ["setuptools >= 34", "wheel"]
```

Example of a `pyproject.toml` file that creates an environment for building [`flit`](https://flit.readthedocs.io/) packages:

```
[build-system]
requires = ["flit"]
```

(Note that this kind of setup is pretty useless without support for [PEP-517](https://www.python.org/dev/peps/pep-0517/).)

Approximate step by step algorithm for supporting both `PEP-517` and "legacy" `./setup.py` deployments (based on PEP-518 & how PIP choose to implement it):

  1. If `skipsdist` in the `tox` section of `tox.ini` is `true`, do nothing
  2. Create a new `virtualenv` to host the `sdist` generation environment
  3. If a file named `pyproject.toml` exists in the `toxinidir`:
     1. Parse the `pyproject.toml` as TOML file (using the python [toml](https://github.com/uiri/toml) library for instance)
     2. Search for the `build-system.requires` key and treat each entry in the array as an [`PEP-508`](https://www.python.org/dev/peps/pep-0508/) dependency specification
     3. If the key was missing or not an array, report an informative error message and exit
     4. Install each found build dependency into the `virtualenv`
 4. Otherwise install `setuptools` and `wheel` into the `virtualenv`
 5. Execute `./setup.py sdist` within the `virtualenv` to generate the `sdist` used for the subsequent testing steps

(The build environment may be reused between subsequent testing sessions, as long as the list of requirements doesn't change.)

### PR DESCRIPTION
**Pull request by gaborbernat**
create a ``.package`` virtual environment to perform build operations inside

@asottile @nicoddemus @obestwalter feel free to take a look at it and try to find holes in it :+1: base implementation there, just need to add documentation and some more tests :100:

FYI @pfmoore (we probably will switch over pip once pip will have pep-517 support and ``build`` command - that being said I find this level a support a bit more generic e.g. python version selection). With this tox should finally allow users to specify their packaging dependencies, and use other than setuptools.

on it

Resolves #573 and #820.

### CODE REVIEW

**Comment by pfmoore** [comment]
Nice! I'll try to find some time to take a proper look at this. (FYI, the initial implementation of PEP 517 in pip won't include a `pip build` command, but `pip wheel` will use PEP 517, I don't know if that's sufficient for you).

**Comment by asottile** [comment]
I'll take a look at this later today -- excited for this!

File: src/tox/config.py [hunk 1]
```diff
@@ -1005,6 +1005,13 @@ def __init__(self, config, inipath):  # noqa
         )
 
         config.skipsdist = reader.getbool("skipsdist", all_develop)
+        config.isolated_build = reader.getbool("isolated_build", False)
+        if config.isolated_build is True:
+            name = ".package"
+            if name not in config.envconfigs:
```
Comment (gaborbernat) [inline src/tox/config.py]: the build env name should be configurable

File: src/tox/package.py [hunk 2]
```diff
@@ -83,3 +99,113 @@ def make_sdist(report, config, session):
...
+    if not previous_config or not previous_config.matches(live_config, True):
+        session.finishvenv(package_venv)
+
+    build_requires = get_build_requires(build_info, package_venv, session)
+    for requirement in build_requires:
+        pkg_requirement = pkg_resources.Requirement(requirement)
+        if pkg_requirement.key not in toml_require:
+            package_venv.envconfig.deps.append(DepConfig(requirement, None))
+
+    if not session.setupenv(package_venv):
```
Comment (gaborbernat) [inline src/tox/package.py]: here we are recreating the env maybe instead just install new dependencies and finalize again the env

File: src/tox/package.py [hunk 3]
```diff
@@ -83,3 +99,113 @@ def make_sdist(report, config, session):
...
+
+    live_config = package_venv._getliveconfig()
+    previous_config = CreationConfig.readconfig(package_venv.path_config)
+    if not previous_config or not previous_config.matches(live_config, True):
+        session.finishvenv(package_venv)
+
+    build_requires = get_build_requires(build_info, package_venv, session)
+    for requirement in build_requires:
+        pkg_requirement = pkg_resources.Requirement(requirement)
+        if pkg_requirement.key not in toml_require:
```
Comment (gaborbernat) [inline src/tox/package.py]: note here we might potentially have a issue that this requirement might differ from the first, maybe check for it and raise if that's the case

File: src/tox/package.py [hunk 4]
```diff
@@ -83,3 +99,113 @@ def make_sdist(report, config, session):
...
+        script = textwrap.dedent(
+            """
+            import sys
+            import {}
+            basename = {}.build_{}("{}", {{ "--global-option": ["--formats=gztar"]}})
+            print(basename)""".format(
+                build_info.backend_module,
+                build_info.backend_object,
+                "sdist",
+                str(config.distdir).replace("\\", "\\\\"),
```
Comment (gaborbernat) [inline src/tox/package.py]: instead of the replace mark the output build directory as a raw string

File: tests/integration/test_package_int.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    assert result.ret == 0, result.out
+
+    result2 = cmd("--sdistonly")
+    assert result2.ret == 0, result.out
+    assert ".package recreate" not in result2.out
+
+
+@pytest.mark.network
+@need_git
+@pytest.mark.skipif(sys.version_info < (3, 0), reason="flit is Python 3 only")
```
Comment (gaborbernat) [inline tests/integration/test_package_int.py]: in CI instead of skip should raise instead

File: tests/lib/__init__.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import subprocess
```
Comment (gaborbernat) [inline tests/lib/__init__.py]: @asottile @nicodeamus stolen this from pip any better way of doing this?

File: tox.ini [hunk 7]
```diff
@@ -125,8 +125,8 @@ multi_line_output = 3
 include_trailing_comma = True
 force_grid_wrap = 0
 line_length = 99
-known_first_party = tox
-known_third_party = apiclient,git,httplib2,oauth2client,packaging,pkg_resources,pluggy,py,pytest,setuptools,six
+known_first_party = tox,tests
+known_third_party = apiclient,git,httplib2,oauth2client,packaging,pkg_resources,pluggy,py,pytest,setuptools,six,tests,toml
```
Comment (gaborbernat) [inline tox.ini]: @asottile for some reason tests shows up as third party even though it is marked as first party

File: src/tox/package.py [hunk 8]
```diff
@@ -83,3 +99,113 @@ def make_sdist(report, config, session):
...
+            )
+        )
+        config.distdir.ensure_dir()
+        result = action.popen([package_venv.envconfig.envpython, "-c", script], returnout=True)
+        return config.distdir.join(result.split("\n")[-2])
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/tox/package.py]: returnout implies universal newlines which means \n only

**Comment by gaborbernat** [comment]
@asottile 

> needs documentation: isolated_build setting, .package environment name
> maybe an example which shows how to configure the executable for the .package environment (and/or other useful things to override there?)

Added some: the config flags plus an example page related to packaging.

> does the .package environment leak into tox --listenvs (or whatever the option is called) -- (either way) should it?
It probably does, we need to filter it out. Will do and add test for it.

> is .package a good enough name to avoid conflicts (probably)? if not maybe .tox-package?
Given that now it's configurable it's good enough I think especially with the ``.`` start to make it private/hidden so to say.

> Will wheel support want to make N .package environments -- does this enable that or make it more difficult?

Wheel support will have to do the ``n`` separate package environments. Going from there maybe we should instead make the isolated_build a choice so people can swap out to other alternatives/options; such as the wheel instead of ``sdist``. I would say the way we have packaging plugabble should be fairly easy to add wheel support later on (no way around it - will have to create up to ``n`` separate build envs though - ``n`` is the number of unique binary type - e.g. 2 if only py2 and py3, but with abi and other parameters can be a lot more).

**Comment by gaborbernat** [comment]
This now is mostly complete. Will merge in on Monday unless we find something significant until then. Will re-read https://www.python.org/dev/peps/pep-0517/ before that to make sure we did not miss anything.

**Comment by obestwalter** [comment]
Better later than never: thank you @gaborbernat for your great work on this :fireworks: :fireworks: :fireworks:

**Pull request merged**